### PR TITLE
`LazyRelay{In,Out}Req` Pooling

### PR DESCRIPTION
**Pull request by rf**
Inspired by Josh's PR, I tried to take a minimal stab at object pooling. This just pools `LazyRelay{In,Out}Req` objects. My pooling implementation also submits stats for all pools.

~~There's still some misplaced frees right now causing test failures, so I need to slog on this a little more to get it right.~~

This now passes all tests.

r @Raynos @jcorbin @kriskowal

### CODE REVIEW

File: lib/object_pool.js [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+ObjectPool.pools = [];
+ObjectPool.timer = null;
+ObjectPool.bootstrapped = false;
+ObjectPool.refs = 0;
+
+ObjectPool.setup = function setup(options) {
+    var pool = new ObjectPool(options);
+    options.Type.alloc = alloc;
+
+    options.Type.prototype.free = function freeThisObj() {
```
Comment (kriskowal) [inline lib/object_pool.js]: Perhaps consider using just one style of closure, like alloc or like free[ThisObj[ect]].

**Comment by kriskowal** [comment]
:eyes: tests. be sure to squash the wip. ping me for another review.

File: lazy_relay.js [hunk 2]
```diff
@@ -344,8 +399,10 @@ function onReadError(err) {
         );
     }
 
+    var conn = self.conn;
     self.onError(err);
-    self.conn.resetAll(err);
+
+    conn.resetAll(err);
```
Comment (jcorbin) [inline lazy_relay.js]: why this change, looks like churn?

File: lazy_relay.js [hunk 3]
```diff
@@ -644,6 +738,7 @@ function handleFrameLazily(frame) {
     frame.setId(self.inreq.id);
     self.inreq.conn.writeToSocket(frame.buffer);
     if (frame.bodyRW.lazy.isFrameTerminal(frame)) {
+        self.alive = false;
```
Comment (jcorbin) [inline lazy_relay.js]: this looks like a prior-existing bug that we didn't mark dead here?

File: lazy_relay.js [hunk 4]
```diff
@@ -523,6 +582,22 @@ function _observeCallReqContFrame(frame) {
...
+    this.remoteAddr = null;
+    this.logger = null;
+    this.inreq = null;
+    this.id = null;
+    this.serviceName = null;
+    this.callerName = null;
+    this.timeout = null;
+    this.operations = null;
+    this.timeHeapHandle = null;
+    this.alive = null;
```
Comment (jcorbin) [inline lazy_relay.js]: is having alive be essentially a "tri-state" intentional? I think it seems useful, but just wanted to check

File: lib/object_pool.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+function ObjectPool(options) {
+    assert(typeof options === 'object', 'expected options object');
+
+    this.Type = options.Type;
+    assert(typeof this.Type === 'function', 'expected options.Type to be constructor function');
+    assert(typeof this.Type.prototype.reset === 'function', 'expected options.Type to have reset method');
+    assert(typeof this.Type.prototype.clear === 'function', 'expected options.Type to have clear method');
+
+    this.maxSize = options.maxSize || 1000;
```
Comment (jcorbin) [inline lib/object_pool.js]: factor out `DEFAULT_MAX_SIZE`

File: lib/object_pool.js (hunk 1 above)
Comment (jcorbin) [inline lib/object_pool.js]: `freeThisObj` could be hoisted to module level fwiw

File: lib/object_pool.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    ObjectPool.timers = options.timers;
+
+    ObjectPool.timer = ObjectPool.timers.setTimeout(
+        ObjectPool.reportStats,
+        ObjectPool.reportInterval
+    );
+};
+
+ObjectPool.unref = function unref() {
+    ObjectPool.refs = Math.min(0, ObjectPool.refs - 1);
```
Comment (jcorbin) [inline lib/object_pool.js]: didn't you mean `Math.max` here?

File: lib/object_pool.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    if (ObjectPool.refs === 0) {
+        ObjectPool.timers.clearTimeout(ObjectPool.timer);
```
Comment (jcorbin) [inline lib/object_pool.js]: may be worth nulling-after-clear for clarity

File: lib/object_pool.js [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+ObjectPool.reportStats = function reportStats() {
+    // Iterate over pools
+    // report their current size
+    var i;
+    var pool;
+    for (i = 0; i < ObjectPool.pools.length; i++) {
+        pool = ObjectPool.pools[i];
+        if (ObjectPool.statsd) {
+            ObjectPool.statsd.gauge(
+                'object-pools.' + pool.name + '.free',
```
Comment (jcorbin) [inline lib/object_pool.js]: all these strings could be constant-by-pool, so maybe we should have the inner part of this loop factored out as say `ObjectPool#reportStats(statsd)`

File: lazy_relay.js (hunk 4 above)
Comment (rf) [inline lazy_relay.js]: Setting everything to `null` is more of a safety thing -- I want to induce reference errors when we accidentally use an object while it's still part of the free list

File: lib/object_pool.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+};
+
+ObjectPool.reportStats = function reportStats() {
+    // Iterate over pools
+    // report their current size
+    var i;
+    var pool;
+    for (i = 0; i < ObjectPool.pools.length; i++) {
+        pool = ObjectPool.pools[i];
+        if (ObjectPool.statsd) {
```
Comment (jcorbin) [inline lib/object_pool.js]: hoist the if out and skip the entire loop

File: lib/object_pool.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+            // This stat is how we infer the maxSize property
+            ObjectPool.statsd.gauge(
+                'object-pools.' + pool.name + '.outstanding',
+                pool.outstanding
+            );
+        }
+    }
+
+    ObjectPool.timer = ObjectPool.timers.setTimeout(
```
Comment (jcorbin) [inline lib/object_pool.js]: do we skip setting a timer too if we have no statsd?

File: lazy_relay.js (hunk 3 above)
Comment (rf) [inline lazy_relay.js]: We didn't have an alive var in `OutReq`s before; I'm not sure we needed it

File: lib/object_pool.js [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    } else {
+        inst = new this.Type();
+        inst._objectPoolIsFreed = false;
+        return inst;
+    }
+};
+
+ObjectPool.prototype.free = function free(inst) {
+    assert(!inst._objectPoolIsFreed, 'object pool double free');
+    inst._objectPoolIsFreed = true;
```
Comment (jcorbin) [inline lib/object_pool.js]: this mutates hidden classes, right?

File: lib/object_pool.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+    inst.clear();
+    if (this.outstanding <= this.maxSize) {
```
Comment (jcorbin) [inline lib/object_pool.js]: I think I want a stat in the contra-case here to know how often I'm blowing my pool limits?

File: lib/object_pool.js [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        }
+    }
+
+    ObjectPool.timer = ObjectPool.timers.setTimeout(
+        ObjectPool.reportStats,
+        ObjectPool.reportInterval
+    );
+};
+
+ObjectPool.prototype.get = function get() {
```
Comment (jcorbin) [inline lib/object_pool.js]: it'd be nice to have hit/miss stats from this

File: lazy_relay.js (hunk 2 above)
Comment (rf) [inline lazy_relay.js]: `LazyRelayInReq#onError` may `free()`, so we need to save a ref to `conn` in order to call `resetAll`.

File: lib/object_pool.js [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    var inst;
+    this.outstanding += 1;
```
Comment (jcorbin) [inline lib/object_pool.js]: isn't `outstanding` wholly redundant (the complement of) `freeList.length`?  Seems like we could just not have it, and the whole class of "can't ever happen" bugs around it de-syncing from `freeList.length`.

File: lib/object_pool.js [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+    if (this.freeList.length) {
+        inst = this.freeList.pop();
+        assert(inst._objectPoolIsFreed, 'instance retreived from pool is free');
+        inst._objectPoolIsFreed = false;
+        return inst;
```
Comment (jcorbin) [inline lib/object_pool.js]: so it's on the caller to call `.reset` right?  And that's necessary because resets are contextual?

File: lib/object_pool.js (hunk 11 above)
Comment (rf) [inline lib/object_pool.js]: It does, but since *all* objects that are of a class setup to be pooled will have this property, it shouldn't have an affect on optimization since the functions will all be compiled against the map for `LazyRelayInReq'`.

File: lib/object_pool.js (hunk 12 above)
Comment (rf) [inline lib/object_pool.js]: You would see outstanding > maxSize, and free !> 0

File: lib/object_pool.js (hunk 14 above)
Comment (rf) [inline lib/object_pool.js]: No, outstanding is *total* instances, including those that couldn't be pooled. It's maximum value infers the maxSize property.

File: lib/object_pool.js (hunk 15 above)
Comment (rf) [inline lib/object_pool.js]: Yes, `reset` is used for it's less-common meaning, a literal *re-*set. Objects will be cleared when `free()`d and `reset` is used to re-initialize.

File: lib/object_pool.js [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+        ObjectPool.timers.clearTimeout(ObjectPool.timer);
+        ObjectPool.timer = null;
+    }
+};
+
+ObjectPool.reportStats = function reportStats() {
+    // Iterate over pools, report their current size
+
+    var i;
+    for (i = 0; i < ObjectPool.pools.length; i++) {
```
Comment (jcorbin) [inline lib/object_pool.js]: what happened to checking for if statsd is a thing?

**Comment by jcorbin** [comment]
lgtm, I'm okay with not having the hit/miss stats initially like this, but think we may want them.

Maybe we could eventually have a `pool.hits` and `pool.misses` number that gets incremented under `.get`, and then reported and reset to zero by `.reportStats` ?

File: lib/object_pool.js (hunk 16 above)
Comment (rf) [inline lib/object_pool.js]: `ObjectPool` can't be bootstrapped without a `statsd` because of the assertions, so the timer will never be set and this function will never be called if there's no `statsd`. The check was actually unnecessary.
